### ISSUE TITLE
Define strides

### ISSUE THREAD

**Issue by aleixalcacer** [FrancescAlted/caterva#49]
Use the concept of strides to simplify the code

**Comment by aleixalcacer** [comment]
Store the strides in the `caterva_array_t` structure to avoid calculating them multiple times in the code. Performance may be improved.

**Comment by muusbolla** [comment]
From the same performance test as [https://github.com/Blosc/caterva/issues/107](https://github.com/Blosc/caterva/issues/107). The test app and caterva/blosc2 libraries were built using gcc -O2 -fPIC -Wall -Wextra -std=gnu11.

I ran "perf record" to get some stack samples. This is another case where small block sizes are really hurt by the application internal implementation. I'm spending 62% of my time doing integer math instead of decompressing things.

Samples: 6K of event 'cycles:u', Event count (approx.): 5568926238
Overhead  Command          Shared Object         Symbol
  46.08%  testapp          testapp               [.] index_unidim_to_multidim
  16.38%  testapp          testapp               [.] index_multidim_to_unidim
  14.86%  testapp          libc.so.6             [.] __memmove_avx_unaligned_erms
  12.81%  testapp          testapp               [.] caterva_copy_buffer
   1.29%  testapp          testapp               [.] ZSTD_safecopy
   1.12%  testapp          [unknown]             [k] 0xffffffff95000ab0
   0.45%  testapp          libc.so.6             [.] __memmove_avx_unaligned
   0.34%  testapp          [unknown]             [k] 0xffffffff947c3877
   0.31%  testapp          testapp               [.] unshuffle8_avx2

   sum(caterva_copy_buffer, index_multidim_to_unidim, index_unidim_to_multidim) = 75.05%
   execution time: 1524ms

**Comment by muusbolla** [comment]
Added some counters to these functions (in the same test case):
index_unidim_to_multidim called 67178496 times
index_multidim_to_unidim called 134287360 times

**Comment by muusbolla** [comment]
Digging in further, index_unidim_to_multidim is called 67108864 times from caterva_copy_buffer and all other calls are only 69632 times. That call site looks like:
```index_unidim_to_multidim(ndim - 1, copy_shape, ncopy, copy_start);```
For my test, I know I'm working with 3 dimensional data and this operates on ndim - 1, so I tried replacing that single call with a call to a function that is an optimized version for 2 dimensions:
```
void index_unidim_to_multidim2(int64_t *shape, int64_t i, int64_t *index) {
    index[0] = i / (shape[1]);
    index[1] = (i % (shape[1]));
}
```
Replacing that one function call site cut the overall execution time of this test from 1500 ms -> 900 ms. This is just an interim hack though - I will try to produce a general solution. Perf trace looks like this now:

  27.19%   testapp  libc.so.6             [.] __memmove_avx_unaligned_erms
  27.17%   testapp   testapp              [.] caterva_copy_buffer
  18.87%   testapp   testapp              [.] index_multidim_to_unidim
   8.46%   testapp   testapp              [.] index_unidim_to_multidim2
   2.28%   testapp   testapp              [.] ZSTD_safecopy
   1.62%   testapp  libc.so.6             [.] __memmove_avx_unaligned
   0.95%   testapp  [unknown]             [k] 0xffffffff95000ab0
   0.56%   testapp  libc.so.6             [.] __memset_avx2_unaligned_erms
   0.54%   testapp   testapp              [.] unshuffle8_avx2

   sum(caterva_copy_buffer, index_multidim_to_unidim, index_unidim_to_multidim2) = 54.5%
   execution time: 900ms

**Comment by muusbolla** [comment]
Final perf report, with the general solution as presented in the pull request. Same test app.
  43.35%  testapp          libc.so.6             [.] __memmove_avx_unaligned_erms
  19.52%  testapp          testapp               [.] copy3dim
   4.81%  testapp          testapp               [.] ZSTD_safecopy
   4.23%  testapp          libc.so.6             [.] __memmove_avx_unaligned
   1.50%  testapp          testapp               [.] unshuffle8_avx2
   1.37%  testapp          [unknown]             [k] 0xffffffff95000ab0
   1.13%  testapp          [unknown]             [k] 0xffffffff947c3877
   1.07%  testapp          libc.so.6             [.] __memset_avx2_unaligned_erms
   1.02%  testapp          testapp               [.] copy_match
   ...
   0.05%  testapp          testapp               [.] caterva_copy_buffer

   sum(caterva_copy_buffer, copy3dim) = 19.57%
   execution time: 553ms

**Comment by FrancescAlted** [comment]
Addressed in #108.
